### PR TITLE
F95 - lowering and bug fixes

### PR DESCRIPTION
**Pull request by schweitzpgi**
- fix freemem bug
  - make pointer/allocatable members emit a TODO
  - add lowering for trivial struct constructor

### CODE REVIEW

File: flang/lib/Lower/ConvertVariable.cpp [hunk 1]
```diff
@@ -1063,9 +1061,6 @@ void Fortran::lower::mapSymbolAttributes(
     }
   };
 
-  if (isHostAssoc)
-    TODO(loc, "host associated");
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertVariable.cpp:1067]: Why are you removing this TODO ? internal procedure are not implemented yet, so it looked legit to me.

File: flang/lib/Lower/IO.cpp [hunk 2]
```diff
@@ -225,11 +225,7 @@ static mlir::FuncOp getOutputFunc(mlir::Location loc,
   // Use descriptors for arrays
   if (auto refTy = type.dyn_cast<fir::ReferenceType>())
     type = refTy.getEleTy();
-  if (type.isa<fir::SequenceType>())
-    return getIORuntimeFunc<mkIOKey(OutputDescriptor)>(loc, builder);
-  // Any unaccounted for types are to be handled here.
-  mlir::emitError(loc, "output for entity type ") << type << " not implemented";
-  TODO(loc, "derived type IO");
```
Comment (jeanPerier) [inline flang/lib/Lower/IO.cpp:232]: Same, here the TODO is still not yet done, this will just crash later.

**Review by jeanPerier — COMMENTED** [review]
Looks good to me but for the host associated / derived type IO TODOs removal that look weird to me.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -2464,9 +2514,10 @@ class ArrayExprLowering {
     auto exv = Fortran::lower::createSomeInitializerExpression(
         loc, converter, toEvExpr(x), symMap, stmtCtx);
     auto val = fir::getBase(exv);
-    mlir::Value mem = builder.create<fir::AllocMemOp>(loc, val.getType());
-    stmtCtx.attachCleanup([=]() { builder.create<fir::FreeMemOp>(loc, mem); });
-    builder.create<fir::StoreOp>(loc, val, mem);
+    auto *bldr = &converter.getFirOpBuilder();
```
Comment (vdonaldson) [inline flang/lib/Lower/ConvertExpr.cpp:2517]: What is the motivation for getting the builder here?

**Comment by mleair** [comment]
LGTM

**Review by mleair — APPROVED** [review]
LGTM

File: flang/lib/Lower/ConvertExpr.cpp (hunk 3 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:2517]: Correctness. We can't forward a copy of a reference sitting on the stack to the future lambda for cleanup.

File: flang/lib/Lower/ConvertVariable.cpp (hunk 1 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertVariable.cpp:1067]: It seems misplaced. This code is determining the properties associated with a variable, which really have nothing to do with whether the variable is host associated or not.

File: flang/lib/Lower/IO.cpp (hunk 2 above)
Comment (schweitzpgi) [reply flang/lib/Lower/IO.cpp:232]: I think "later" here is fine and caught with a message about no code to lower to LLVM. There is no reason then to catch this as early as possible. Removing the TODO exposes other problems, like the TODO in codegen.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 3 above)
Comment (vdonaldson) [reply flang/lib/Lower/ConvertExpr.cpp:2517]: Thanks for the explanation.

**Review by vdonaldson — APPROVED** [review]

**Pull request merged**